lighthouse-core/config/config.js

```diff
@@ -172,32 +199,73 @@ function merge(base, extension, overwriteArrays = false) {
     return merged;
   } else if (typeof extension === 'object') {
     if (typeof base !== 'object') throw new TypeError(`Expected object but got ${typeof base}`);
+    if (Array.isArray(base)) throw new TypeError('Expected object but got Array');
     Object.keys(extension).forEach(key => {
       const localOverwriteArrays = overwriteArrays ||
         (key === 'settings' && typeof base[key] === 'object');
-      base[key] = merge(base[key], extension[key], localOverwriteArrays);
+      base[key] = _merge(base[key], extension[key], localOverwriteArrays);
     });
     return base;
   }
 
   return extension;
 }
 
+/**
+ * Until support of jsdoc templates with constraints, type in config.d.ts.
+ * See https://github.com/Microsoft/TypeScript/issues/24283
+ * @type {LH.Config.Merge}
+ */
+const merge = _merge;
+
+/**
+ * @template T
+ * @param {Array<T>} array
+ * @return {Array<T>}
+ */
 function cloneArrayWithPluginSafety(array) {
   return array.map(item => {
-    return typeof item === 'object' ? Object.assign({}, item) : item;
+    if (typeof item === 'object') {
+      // Return copy of instance and prototype chain (in case item is instantiated class).
+      return Object.assign(
+        Object.create(
+          Object.getPrototypeOf(item)
+        ),
+        item
+      );
+    }
+
+    return item;
   });
 }
 
+/**
+ * // TODO(bckenny): could adopt "jsonified" type to ensure T will survive JSON
+ * round trip: https://github.com/Microsoft/TypeScript/issues/21838
+ * @template T
+ * @param {T} json
+ * @return {T}
+ */
 function deepClone(json) {
-  const cloned = JSON.parse(JSON.stringify(json));
+  return JSON.parse(JSON.stringify(json));
+}
+
+/**
+ * Deep clone a ConfigJson, copying over any "live" gatherer or audit that
+ * wouldn't make the JSON round trip.
+ * @param {LH.Config.Json} json
+ * @return {LH.Config.Json}
+ */
+function deepCloneConfigJson(json) {
+  const cloned = deepClone(json);
 
   // Copy arrays that could contain plugins to allow for programmatic
   // injection of plugins.
-  if (Array.isArray(json.passes)) {
-    cloned.passes.forEach((pass, i) => {
+  if (Array.isArray(cloned.passes) && Array.isArray(json.passes)) {
```

Comment:
could this ever not be true if one of them is true? just for TS?

Reply:
> just for TS?

just for TS and pretty annoying :)